drivers/serial/uart_cmsdk_apb.c

```diff
@@ -414,13 +426,13 @@ void uart_cmsdk_apb_isr(void *arg)
 	volatile struct uart_cmsdk_apb *uart = UART_STRUCT(dev);
 	struct uart_cmsdk_apb_dev_data *data = DEV_DATA(dev);
 
-	/* Clear pending interrupts */
-	uart->intclear = UART_RX_IN | UART_TX_IN;
-
 	/* Verify if the callback has been registered */
 	if (data->irq_cb) {
 		data->irq_cb(data->irq_cb_data);
 	}
+
+	/* Clear pending interrupts */
+	uart->intclear = UART_RX_IN | UART_TX_IN;
```

Comment:
@pfalcon Why are we clearing the pending interrupt after calling the callback?
We do not have control on how much time the callback will complete, hence to not loose events we should clear the interrupt first.

Reply:
> We do not have control on how much time the callback will complete

Because ISR is *required* to check status of interrupts (that's the API contract). So, obviously, we cannot clear them before it does. Well, the API is also prepared to handle that case too. We can cache interrupt status, acknowledge *only* the interrupts set there, then make ISR look into cached status. It's still not exactly clear if there maybe races, depending on the exact interrupt behavior. 

Otherwise, how we work is by gradual improvements, solving an issue at hand. And this patch is a tangible improvement to the old behavior of the driver, which didn't work at all. Well, now need to fix this race.

Reply:
I am not sure I fully understand. Let's try to provide some context to my previous comment:
- Here you are inside the ISR (uart_cmsdk_apb_isr) and I do not think we are breaking any contract clearing the interrupt before the user callback of which as I said in my previous comment we do not have control since it is defined by the application.
- Furthermore, intclear is a register of the specific IP (CMSDK UART in this case) of which the application is not supposed to know because of the uart level (include/uart.h) of abstraction defined by Zephyr.

Unless you have a specific reason, we should definitely move this lines back to their original position. If they are breaking any usecase, please provide the usecase and I am happy to have a look into it.

Reply:
> Ok, let's clarify the terminology ;-). uart_cmsdk_apb_isr() is nothing but _ISR dispatcher_. Real interrupt handling happens in _user-installed ISR callback_, or just "ISR" for short.

@pfalcon Ok, Thanks for providing the context, now I think I understand a bit more of what you want to do here. BTW, by definition uart_cmsdk_apb_isr() is an ISR and it is the only thing under the domain of the driver. 

How about we organize things in this way:
uart_cmsdk_apb_isr() {
  Check the reason
  Clear the interrupt
  Call the user CB passing the reason
}

This should be compliant with the contract you are defining, safer (because independent from the application) and should even simplify the way on which the user callback is written?

Please let me know your thoughts.

Reply:
@fvincenzo

> Call the user CB passing the reason

Unfortunately, that's effectively an API break. I love (making) API breaks - what can be better than breaking an API once in the morning once in a while? ;-) Unfortunately, users don't like API breaks :-(. And we breaks APIs in Zephyr all the time (because, fairly speaking, a lot of APIs in Zephyr suck). That means we should approach each case carefully, and improve carefulness over time. Well, I know that you know all that. Just to explain why I would say "no" in this case (see also previous case carefully approached to minimize breakage effects: 57286afdd60bc0522aa09848f4862e930c75eb7d). And the main reason is existing API offers a way to do that already, exactly the one I described in the first comment in this thread: https://github.com/zephyrproject-rtos/zephyr/pull/11314#discussion_r234300777 .

So, let me prepare a patch over this week for you to review. And as I mentioned, another reason not to try to "improve" existing API too much is because a new one is coming: https://github.com/zephyrproject-rtos/zephyr/pull/10820 . So, what we should is to make sure this new API is definitely better than the old one, and doesn't have (obvious) goofs.

Reply:
Ok, I will wait for your patch to comment further. And thanks for pointing out the new API, I will follow that as well.

Just to be clear: I am not asking for an API breakage here. What I was suggesting does not involve that. I am against any API breakage that violates the Zephyr versioning contract.

Said that, "Call the user CB passing the reason" can lead to a misunderstanding. What I meant here is passing the reason via the existin API, in fact the user callback is supposed to read the reason using the existin API. You have to manage IRQ locking a bit but it can be achieved.

Reply:
> Just to be clear: I am not asking for an API breakage here.
> What I meant here is passing the reason via the existin API, in fact the user callback is supposed to read the reason using the existin API.

Ah, that clarifies it, and we're on the same line then. These 2 API functions are  uart_irq_rx_ready() and uart_irq_tx_ready(), and the whole ISR callback looks as in https://github.com/zephyrproject-rtos/zephyr/pull/11314#issuecomment-438267355 . And the change we seek is that those 2 funcs operate on a capture cached copy of hw INT status register, while the register itself has those 2 bits acknowledges ASAP.